src/pages/test-finalisation/pass-finalisation/pass-finalisation.ts

```diff
@@ -260,6 +265,70 @@ export class PassFinalisationPage extends PracticeableBasePageComponent {
     return ctrlHasErrors && this.form.get(this.passCertificateCtrl).dirty;
   }
 
+  // Custom validator for FormControls
+  validatePassCertificate(c: FormControl) {
+    return this.isPassCertificateValid(c.value) ? null :
+      {
+        validatePassCertificate: {
+          valid: false,
+        },
+      };
+  }
+
+  isPassCertificateValid(certificate: string | null): boolean {
+    if (!certificate || certificate.length !== 8) {
+      return false;
+    }
+    if (!this.isLetter(certificate[0])) {
+      return false;
+    }
+
+    const checkDigit = this.calculateMod37CheckDigit(certificate);
+
+    if (checkDigit === 'invalid') {
+      return false;
+    }
+
+    // compare entered checkdigit with generated one
+    if (checkDigit !== certificate.toUpperCase()[7]) {
+      return false;
+    }
+
+    return true;
+  }
+
+  isLetter(char: string): boolean {
+    return char.length === 1 && char.match(/[a-z]/i) !== null;
+  }
+
+  calculateMod37CheckDigit(certificate: string): string {
```

Comment:
adjust the type-hinting since the function returns bool or a string

Reply:
it actually only returns string - either the checkDigit or the string 'invalid'.... no bool.